include/sound/sof/topology.h

```diff
@@ -147,6 +148,20 @@ struct sof_ipc_comp_src {
 	uint32_t rate_mask;	/**< SOF_RATE_ supported rates */
 } __packed;
 
+/* generic ASRC component */
+struct sof_ipc_comp_asrc {
+	struct sof_ipc_comp comp;
+	struct sof_ipc_comp_config config;
+	/* either source or sink rate must be non zero */
+	uint32_t source_rate;		/**< source rate or 0 for variable */
+	uint32_t sink_rate;		/**< sink rate or 0 for variable */
+	uint32_t asynchronous_mode;	/**< synchronous 0, asynchronous 1 */
+	uint32_t operation_mode;	/**< push 0, pull 1 */
```

Comment:
likewise isn't this determined by the source_rate or sink_rate being 0?

Reply:
In SRC there's some logic to decide push/pull internally but it remains fuzzy due to no capability to process desired amount of samples with granularity of one like the ASRC can do.

I had the logic in beginning similarly for ASRC too but I decided to expose this control to IPC to not restrict flexibility with IPC. Though very likely setting will be most of the time push for playback and pull for capture if we want to operate with constant # of frames on FW side and we assume ASRC is placed next to DAI in topology.

The topologies that I've prepared (no PR yet) have fixed setting for playback/capture but it could be changed easily when it's exposed to IPC and tokens.

Reply:
source_rate == 0 -> pull
sink_rate == 0 -> push

Even if I confused push/pull, don't you have a 1:1 relationship between the which of the two rates is zero and the operation mode? I just don't get why we would pass redundant information.

Reply:
Yes I agree the controls you propose are very likely all that is needed. I'll think this more and revert to "operation_mode" removal if I can't think of use cases where we need the extra control.

If we decide that ASRC needs to be always next to a DAI it reduces a lot possible use scenarios. Though not sure if it impacts control needs

Also in many usages the rates for source and sink are fixed but nothing prevents to set one to zero to control push/pull and retrieve the actual from stream.

Reply:
well the rate 0 is kind of odd to be honest. In a lot of cases we have a nominal rate that is e.g. 48kHz, but it's not exactly 48kHz. It's likely to help the adjustment routine if the initial values are set.

Maybe we need to remove this usage, and use the operational mode to specific which of the two is nominal/actual. You could even have cases where both rates can vary.

Reply:
I have improved the descriptions to the version that will be pushed next. I think this topology / component new() IPC is quite sufficient now. We can add run-time control command later to adjust the conversion factor from user space.

The FW currently finds the DAI to track from downstream in push mode and from upstream in pull mode. It assumes one side of ASRC is synchronous to FW wall clock. But it could check both sides by default as well. So supporting a loop between two slave interfaces within FW should be OK with this IPC version (but need more work on FW side vs. what I have now).